pil-analyzer/src/pil_analyzer.rs

```diff
@@ -174,6 +176,71 @@ impl PILAnalyzer {
         }
     }
 
+    pub fn match_exhaustiveness_check(&self) {
+        for (name, (_, value)) in &self.definitions {
+            let Some(value) = value else { continue };
+            let patterns = match value {
+                FunctionValueDefinition::Expression(TypedExpression { e, .. }) => {
+                    if let Expression::MatchExpression(_, arms) = e {
+                        arms.iter()
+                            .map(|arm| arm.pattern.clone())
+                            .collect::<Vec<_>>()
+                    } else {
+                        continue;
+                    }
+                }
+                _ => continue,
+            };
+            let witnesses = Self::usefulness(&patterns, &Pattern::CatchAll);
+            if witnesses.is_empty() {
+                panic!("Function {name} is not exhaustive")
+            }
+        }
+    }
+
+    pub fn usefulness(patterns: &[Pattern], new_pattern: &Pattern) -> Vec<Pattern> {
+        let mut witnesses = Vec::new();
+
+        if patterns.is_empty() || patterns.iter().any(|p| p.is_irrefutable()) {
+            return vec![];
+        }
+
+        let mut expanded_patterns = patterns.to_vec();
+        expanded_patterns.push(new_pattern.clone());
+        let constructors = expanded_patterns.clone(); //self.constructors_from_patterns(expanded_patterns.as_slice());
+
+        for constructor in &constructors {
+            let specialized_new_pattern = new_pattern.specialize(constructor);
```

Comment:
Doesn't specialize need to return both a new constructor and a new pattern?

Reply:
No, specialize only "removes a layer" of the pattern if the constructor is the same. 

```
specialize(pattern = [1,2,3], cons = [_,_,_,_]) => (1 ,2 ,3) 
```
because they are both `[]`.

also unspecialize(...) is literally specialize(...)<sup>-1</sup> so
```
unspecialize(pattern = (1,2,3), cons = [_,_,_,_] ) => [1,2,3]
```